File: sdk/flutter/praxy_core/test/support/fake_transport.dart
import 'dart:convert';

import 'package:praxy_core/praxy_core.dart';

typedef ResponseBuilder = TransportResponse Function(TransportRequest request);

/// A [Transport] test double: replays canned responses (or a handler) instead of
/// hitting the network, and records every request sent through it for assertions.
final class FakeTransport implements Transport {
  FakeTransport(this._handler);

  final ResponseBuilder _handler;
  final List<TransportRequest> requests = [];

  @override
  Future<TransportResponse> send(TransportRequest request) async {
    requests.add(request);
    return _handler(request);
  }

  @override
  void close() {}
}

TransportResponse jsonResponse(int status, Object? body, {Map<String, String> headers = const {}}) {
  final bytes = body == null ? const <int>[] : utf8.encode(jsonEncode(body));
  return TransportResponse(statusCode: status, headers: headers, bodyBytes: bytes);
}

TransportResponse emptyResponse(int status, {Map<String, String> headers = const {}}) =>
    TransportResponse(statusCode: status, headers: headers, bodyBytes: const []);

File: sdk/js/packages/core/test/support/fake-transport.ts
import type { Transport, TransportRequest, TransportResponse } from "../../src/transport";

export type ResponseBuilder = (request: TransportRequest) => TransportResponse;

/**
 * A `Transport` test double: replays a handler's response instead of hitting the network, and
 * records every request sent through it for assertions. Mirrors `praxy_core`'s
 * `test/support/fake_transport.dart`.
 */
export class FakeTransport implements Transport {
  readonly requests: TransportRequest[] = [];

  constructor(private readonly handler: ResponseBuilder) {}

  async send(request: TransportRequest): Promise<TransportResponse> {
    this.requests.push(request);
    return this.handler(request);
  }
}

export function jsonResponse(status: number, body: unknown, headers: Record<string, string> = {}): TransportResponse {
  return { status, headers, body: body === undefined ? "" : JSON.stringify(body) };
}

export function emptyResponse(status: number, headers: Record<string, string> = {}): TransportResponse {
  return { status, headers, body: "" };
}

/** A binary body, as `FetchTransport` returns one for a request that asked for `expect: "bytes"`. */
export function bytesResponse(
  status: number,
  bytes: Uint8Array,
  headers: Record<string, string> = {},
): TransportResponse {
  return { status, headers, body: "", bodyBytes: bytes };
}
